### PR TITLE
Driver filter search

### PR DESCRIPTION
**Pull request by alavieille**
https://github.com/open-orchestra/open-orchestra-cms-bundle/pull/1053
https://github.com/open-orchestra/open-orchestra-workflow-function-bundle/pull/49
https://github.com/open-orchestra/open-orchestra-libs/pull/23#

### CODE REVIEW

File: Mapping/Metadata/Driver/AnnotationDriver.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    /**
+     * @param \ReflectionClass $class
+     *
+     * @return \Metadata\ClassMetadata|null
+     */
+    public function loadMetadataForClass(\ReflectionClass $class)
+    {
+
```
Comment (ngilain) [inline Mapping/Metadata/Driver/AnnotationDriver.php]: useless empty line ?

File: Mapping/Metadata/Driver/AnnotationDriver.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+     * @param AnnotationReader $reader
+     * @param string           $propertySearchMetadataClass
+     * @param string           $mergeableClassMetadata
+     * @param string           $annotationClass
+     */
+    public function __construct(AnnotationReader $reader, $propertySearchMetadataClass, $mergeableClassMetadata, $annotationClass)
+    {
+        $this->reader = $reader;
+        $this->propertySearchMetadataClass = $propertySearchMetadataClass;
+        $this->mergeableClassMetadata = $mergeableClassMetadata;
```
Comment (ngilain) [inline Mapping/Metadata/Driver/AnnotationDriver.php]: Can you check that the class exists and eventually respects an interface ?

File: Mapping/Metadata/Driver/AnnotationDriver.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            $propertyMetadata = new $this->propertySearchMetadataClass($class->getName(), $reflectionProperty->getName());
+
+            $annotations = $this->reader->getPropertyAnnotations(
+                $reflectionProperty,
+                $this->annotationClass
+            );
+
+            if (!empty($annotations)) {
+                $existAnnotation = true;
+                foreach($annotations as $annotation) {
```
Comment (ngilain) [inline Mapping/Metadata/Driver/AnnotationDriver.php]: space between foreach and ( ?

File: Mapping/Metadata/Driver/AnnotationDriver.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                        }
+                        $classMetadata->addPropertyMetadata($propertyMetadata);
+                    }
+                }
+            }
+        }
+
+        return (true === $existAnnotation) ? $classMetadata : null;
+    }
+
```
Comment (ngilain) [inline Mapping/Metadata/Driver/AnnotationDriver.php]: useless empty space ?

File: Mapping/Metadata/Driver/XmlDriver.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * @param FileLocatorInterface $locator
+     * @param string               $propertySearchMetadataClass
+     * @param string               $mergeableClassMetadata
+     */
+    public function __construct(FileLocatorInterface $locator, $propertySearchMetadataClass, $mergeableClassMetadata)
+    {
+        parent::__construct($locator);
+        $this->propertySearchMetadataClass = $propertySearchMetadataClass;
```
Comment (ngilain) [inline Mapping/Metadata/Driver/XmlDriver.php]: You should check that the class exists here too

File: Mapping/Metadata/Driver/XmlDriver.php [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            return $classMetadata;
+        }
+
+        return null;
+    }
+
+    /**
+     * @param string $key
+     *
+     * @return array|
```
Comment (ngilain) [inline Mapping/Metadata/Driver/XmlDriver.php]: Pipe useless or missing type/value

File: Mapping/Metadata/Driver/XmlDriver.php [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+     */
+    protected function extractKey($key)
+    {
+        $key = array_map('trim', explode(',', $key));
+        if (count($key) == 1){
```
Comment (ngilain) [inline Mapping/Metadata/Driver/XmlDriver.php]: space between ) & {

File: Mapping/Metadata/Driver/YamlDriver.php [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * @param FileLocatorInterface $locator
+     * @param string               $propertySearchMetadataClass
+     * @param string               $mergeableClassMetadata
+     */
+    public function __construct(FileLocatorInterface $locator, $propertySearchMetadataClass, $mergeableClassMetadata)
+    {
+        parent::__construct($locator);
+        $this->propertySearchMetadataClass = $propertySearchMetadataClass;
```
Comment (ngilain) [inline Mapping/Metadata/Driver/YamlDriver.php]: Check again here

File: Mapping/Tests/Mapping/Metadata/Driver/YamlDriverTest.php [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+ */
+class YamlDriverTest extends AbstractDriverTest
+{
+    /**
+     * Set Up
+     */
+    public function setUp()
+    {
+        $dirs = array('OpenOrchestra\Mapping\Tests\Mapping\Metadata\Driver\FakeClass' => __DIR__ . '/yml');
+        $fileLocaltor = new FileLocator($dirs);
```
Comment (ngilain) [inline Mapping/Tests/Mapping/Metadata/Driver/YamlDriverTest.php]: fileLocaltor :p

File: MongoBundle/DependencyInjection/OpenOrchestraMongoExtension.php [hunk 10]
```diff
@@ -23,7 +23,41 @@ public function load(array $configs, ContainerBuilder $container)
...
         $loader = new Loader\YamlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
-        $loader->load('annotation_reader.yml');
+        $loader->load('metadata_search_reader.yml');
         $loader->load('filter_strategy.yml');
+
+        //cache metadata
+        $cacheDirectory = $config['search_metadata']['cache_dir'];
+        $cacheDirectory = $container->getParameterBag()->resolveValue($cacheDirectory);
+        if (!is_dir($cacheDirectory)) {
+            mkdir($cacheDirectory, 0777, true);
```
Comment (ngilain) [inline MongoBundle/DependencyInjection/OpenOrchestraMongoExtension.php]: can you restrict the right ? 777 is too much

**Comment by alavieille** [comment]
Pr updated @itkg-ngilain

**Comment by ngilain** [comment]
:+1:

**Pull request merged**